libpages/root.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libpages
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"strings"
+
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/kbfs/libfs"
+	"github.com/keybase/kbfs/libkbfs"
+	"github.com/keybase/kbfs/tlf"
+	"go.uber.org/zap"
+)
+
+// ErrInvalidKeybasePagesRecord is returned when the kbp= DNS record for a
+// domain is invalid.
+type ErrInvalidKeybasePagesRecord struct{}
+
+// Errror returns the error interface.
+func (ErrInvalidKeybasePagesRecord) Error() string {
+	return "invalid TXT record"
+}
+
+// RootType specifies the type of a root.
+type RootType int
+
+const (
+	_ RootType = iota
+	// KBFSRoot means the root is backed by a KBFS path.
+	KBFSRoot
+	// GitRoot means the root is backed by a git repo stored in KBFS.
+	GitRoot
+)
+
+// String implements the fmt.Stringer interface
+func (t RootType) String() string {
+	switch t {
+	case KBFSRoot:
+		return "kbfs"
+	case GitRoot:
+		return "git"
+	default:
+		return "unknown"
+	}
+}
+
+// Root defines the root of a static site hosted by Keybase Pages.
+type Root struct {
+	Type            RootType
+	TlfType         tlf.Type
+	TlfNameUnparsed string
+	PathUnparsed    string
+}
+
+// MakeFS makes a http.FileSystem from *r, to be used by http package to serve
+// through HTTP.
+func (r *Root) MakeFS(ctx context.Context, log *zap.Logger,
+	kbfsConfig libkbfs.Config) (hfs http.FileSystem, err error) {
+	defer func() {
+		if err == nil {
+			log.Info("root.MakeFS",
+				zap.String("type", r.Type.String()),
```

Comment:
Might be nice to construct a slice of these zap strings outside the `if` statement, to avoid duplication.